File: apps/mobile/android/src/main/kotlin/com/messenger/crypto/SenderKey.kt
// apps/mobile/android/src/main/kotlin/com/messenger/crypto/SenderKey.kt
package com.messenger.crypto

import com.goterl.lazysodium.LazySodium
import com.goterl.lazysodium.interfaces.SecretBox

class SenderKey(private val sodium: LazySodium) {

    companion object {
        private const val NONCEBYTES = SecretBox.NONCEBYTES
        private const val MACBYTES   = SecretBox.MACBYTES
    }

    fun encrypt(plaintext: ByteArray, senderKey: ByteArray): Pair<ByteArray, ByteArray> {
        require(senderKey.size == 32) { "senderKey must be 32 bytes, got ${senderKey.size}" }
        val nonce = sodium.randomBytesBuf(NONCEBYTES)
        return encryptWithNonce(plaintext, senderKey, nonce) to nonce
    }

    fun encryptWithNonce(plaintext: ByteArray, senderKey: ByteArray, nonce: ByteArray): ByteArray {
        require(senderKey.size == 32) { "senderKey must be 32 bytes, got ${senderKey.size}" }
        require(nonce.size == NONCEBYTES) { "nonce must be $NONCEBYTES bytes, got ${nonce.size}" }
        val ciphertext = ByteArray(plaintext.size + MACBYTES)
        check(sodium.cryptoSecretBoxEasy(ciphertext, plaintext, plaintext.size.toLong(), nonce, senderKey)) {
            "SenderKey encrypt failed"
        }
        return ciphertext
    }

    fun decrypt(ciphertext: ByteArray, nonce: ByteArray, senderKey: ByteArray): ByteArray {
        require(senderKey.size == 32) { "senderKey must be 32 bytes, got ${senderKey.size}" }
        require(nonce.size == NONCEBYTES) { "nonce must be $NONCEBYTES bytes, got ${nonce.size}" }
        require(ciphertext.size >= MACBYTES) { "ciphertext too short" }
        val plaintext = ByteArray(ciphertext.size - MACBYTES)
        check(sodium.cryptoSecretBoxOpenEasy(plaintext, ciphertext, ciphertext.size.toLong(), nonce, senderKey)) {
            "SenderKey decrypt failed — wrong key or corrupted ciphertext"
        }
        return plaintext
    }
}

File: apps/mobile/ios/Sources/MessengerCrypto/SenderKey.swift
// SenderKey.swift — Sender Key шифрование для групповых чатов.
// Совместим с Kotlin-реализацией (Desktop/Android) и TypeScript-реализацией (web client).
//
// Алгоритм: crypto_secretbox_easy (XSalsa20-Poly1305), тот же примитив, что и Ratchet.

import Sodium

enum SenderKeyError: Error {
    case encryptFailed
    case decryptFailed
    case invalidInput(String)
}

final class SenderKey {
    private let sodium = Sodium()

    /// Шифрует plaintext с помощью senderKey, генерирует случайный nonce.
    /// Возвращает (ciphertext, nonce).
    func encrypt(plaintext: Bytes, senderKey: Bytes) throws -> (ciphertext: Bytes, nonce: Bytes) {
        guard senderKey.count == 32 else {
            throw SenderKeyError.invalidInput("senderKey must be 32 bytes, got \(senderKey.count)")
        }
        let nonce = sodium.secretBox.nonce()
        let ciphertext = try encryptWithNonce(plaintext: plaintext, senderKey: senderKey, nonce: nonce)
        return (ciphertext, nonce)
    }

    func encryptWithNonce(plaintext: Bytes, senderKey: Bytes, nonce: Bytes) throws -> Bytes {
        guard senderKey.count == 32 else {
            throw SenderKeyError.invalidInput("senderKey must be 32 bytes, got \(senderKey.count)")
        }
        guard nonce.count == sodium.secretBox.NonceBytes else {
            throw SenderKeyError.invalidInput("nonce must be \(sodium.secretBox.NonceBytes) bytes, got \(nonce.count)")
        }
        guard let ciphertext = sodium.secretBox.seal(message: plaintext, secretKey: senderKey, nonce: nonce) else {
            throw SenderKeyError.encryptFailed
        }
        return ciphertext
    }

    /// Расшифровывает ciphertext.
    func decrypt(ciphertext: Bytes, nonce: Bytes, senderKey: Bytes) throws -> Bytes {
        guard senderKey.count == 32 else {
            throw SenderKeyError.invalidInput("senderKey must be 32 bytes, got \(senderKey.count)")
        }
        guard nonce.count == sodium.secretBox.NonceBytes else {
            throw SenderKeyError.invalidInput("nonce must be \(sodium.secretBox.NonceBytes) bytes, got \(nonce.count)")
        }
        guard ciphertext.count >= sodium.secretBox.MacBytes else {
            throw SenderKeyError.invalidInput("ciphertext too short")
        }
        guard let plaintext = sodium.secretBox.open(authenticatedCipherText: ciphertext, secretKey: senderKey, nonce: nonce) else {
            throw SenderKeyError.decryptFailed
        }
        return plaintext
    }
}
